test/common/json/json_fuzz_test.cc

```diff
@@ -32,12 +32,13 @@ DEFINE_FUZZER(const uint8_t* buf, size_t len) {
 
     // MessageUtil::getYamlStringFromMessage automatically convert types, so we have to do another
     // round-trip.
-    std::string yaml = MessageUtil::getYamlStringFromMessage(message);
+    std::string yaml = MessageUtil::getYamlStringFromMessage(message, true, false, true);
     ProtobufWkt::Struct yaml_message;
     TestUtility::loadFromYaml(yaml, yaml_message);
 
     ProtobufWkt::Struct message3;
-    TestUtility::loadFromYaml(MessageUtil::getYamlStringFromMessage(yaml_message), message3);
+    TestUtility::loadFromYaml(
+        MessageUtil::getYamlStringFromMessage(yaml_message, true, false, true), message3);
```

Comment:
Hi, may be we can keep this test unchanged, and add a new test for your new case?

Reply:
Well, not changing the test will prevent it from passing for certain fuzz input (e.g. input that has tabs in the values or keys). So how to do this without changing the test? Reject fuzz input, that does not suit the test? I am open for ideas!

Reply:
Then, how the test passed before this PR? 🤔 If you think it's some coincidences, and add quotes is the right option always, then may be we needn't `quote_yaml_strings` parameter. We can add quotes just by default.

Reply:
I don't know. I can only assume that the fuzzer hasn't used the offending character(s) before. 

> We can add quotes just by default.
I've thought of that, too. I wanted the code change as small as possible and as optional as possible. I am quite new to envoy and do not overlook the impacts of changes at all yet. So if you vote that the option should be removed, then I am happy to comply. Any thoughts?

Reply:
🤔 We should prove that the current implementation has this problem in the first. And then we can discuss the details of this patch.

Reply:
With commit 6b5ec04d8f99a15365f77bbb3c72b3181e940ea7 I have added a testcase to the corpus, that shows the test failing. Only after adding above changes (now in commit eaf077e19f81bfee62753a5b89198c22339e0af3) the test passes.

Sorry, I should have added the test input straight from the beginning.

Reply:
May a simple unit test would be better for this scene? cc @adisuissa

Reply:
@wbpcode I'm not sure I fully grok your question.

This code is part of the fuzz testing code.
The fuzzer essentially spits out random strings that are parsed using the JSON parser.
The test then goes through converting to YAML and back to proto to validate contents equality. This PR converts the parsed JSON into YAML with quotes around the strings.

Do you think there's a better way to handle the fuzzer input?

Reply:
> Some strings in a protobuf message may not be valid in YAML,
like strings containing tabs.

I mean should we add a unit test to ensure this problem is exist first? Then if it ok, we can make the quotes adding to be default behaviour and no need a new function parameter `quote_yaml_strings`.

Reply:
Mhh, the test input in commit 6b5ec04d8f99a15365f77bbb3c72b3181e940ea7 does not show you, that there is an issue? What more do you like to see?

Of course can we add a unit test. That would be exactly like the input in commit 6b5ec04d8f99a15365f77bbb3c72b3181e940ea7 adding more code to maintain to an already not small code base. What would the benefit be?

Whether the additional parameter makes sense or it should be defaulted can not be decided by me. I am too new in this project.

Reply:
@vehre-x41 Sorry. But i did not find the failure CI. Do you run this test in your local env? Can you post your failure logs?

Reply:
To replay the error:
`
$ git reset 6b5ec04d8f99a15365f77bbb3c72b3181e940ea7
$ bazel build -c dbg test/common/json/json_fuzz_test
Starting local Bazel server and connecting to it...
INFO: Analyzed target //test/common/json:json_fuzz_test (360 packages loaded, 17511 targets configured).
INFO: Found 1 target...
Target //test/common/json:json_fuzz_test up-to-date:
  bazel-bin/test/common/json/json_fuzz_test
INFO: Elapsed time: 1253.410s, Critical Path: 90.30s
INFO: 1342 processes: 27 internal, 1315 linux-sandbox.
INFO: Build completed successfully, 1342 total actions
$ bazel-bin/test/common/json/json_fuzz_test test/common/json/json_corpus/clusterfuzz-testcase-minimized-json_fuzz_test-4779322450903040.fuzz 
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from CorpusExamples/FuzzerCorpusTest
[ RUN      ] CorpusExamples/FuzzerCorpusTest.RunOneCorpusFile/0
[2022-03-29 13:14:04.088][170894][info][misc] [test/fuzz/main.cc:47] Corpus file: test/common/json/json_corpus/clusterfuzz-testcase-minimized-json_fuzz_test-4779322450903040.fuzz
[2022-03-29 13:14:04.091][170894][critical][assert] [test/common/json/json_fuzz_test.cc:41] assert failure: TestUtility::protoEqual(yaml_message, message3).
[2022-03-29 13:14:04.091][170894][critical][backtrace] [./source/server/backtrace.h:104] Caught Aborted, suspect faulting address 0x3e800029b8e
[2022-03-29 13:14:04.091][170894][critical][backtrace] [./source/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
[2022-03-29 13:14:04.091][170894][critical][backtrace] [./source/server/backtrace.h:92] Envoy version: 0/1.22.0-dev/test/DEBUG/BoringSSL
[2022-03-29 13:14:04.107][170894][critical][backtrace] [./source/server/backtrace.h:96] #0: Envoy::SignalAction::sigHandler() [0x2f305f8]
[2022-03-29 13:14:04.109][170894][critical][backtrace] [./source/server/backtrace.h:96] #1: __restore_rt [0x7f3980290750]
`

This is what I do and the error I get. Note, this is a fuzz test, not a regular unit test.

Reply:
I see. But I still cannot reproduce it with the latest envoy commit. 🤣 

```
envoybuild@742e6f49803c:/source$ bazel build -c dbg test/common/json/json_fuzz_test
envoybuild@742e6f49803c:/source$ bazel-bin/test/common/json/json_fuzz_test ./clusterfuzz-testcase-minimized-json_fuzz_test-4779322450903040.fuzz 
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from CorpusExamples/FuzzerCorpusTest
[ RUN      ] CorpusExamples/FuzzerCorpusTest.RunOneCorpusFile/0
[2022-03-29 13:04:14.971][37609][info][misc] [test/fuzz/main.cc:47] Corpus file: ./clusterfuzz-testcase-minimized-json_fuzz_test-4779322450903040.fuzz
[       OK ] CorpusExamples/FuzzerCorpusTest.RunOneCorpusFile/0 (5 ms)
[----------] 1 test from CorpusExamples/FuzzerCorpusTest (5 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (5 ms total)
[  PASSED  ] 1 test.
```

Reply:
@vehre-x41 **Hi, thanks for calling this out. And thanks for your patience.** I replayed this error by download your fuzz file (I am not sure why, but copy is not work). 

Then, for always get valid message, IMO, we can set `YAML::EMITTER_MANIP::DoubleQuoted` option by default. 

cc @adisuissa

Reply:
A few other things: YAML automatically adds double quotes to strings containing special characters, but \r doesn't seem to be handled. This may be the cause of the problem.

Reply:
You're welcome.

Well, the convert routine `getYamlStringFromMessage()` is using `getJsonStringFromMessage()` and because in json a tab is valid in a name (Json quotes all ids/strings by default) there is no such problem. Therefore the `getJsonStringFromMessage()` sees no reason to extra quote or escape a string containing tabs. So we found a small difference between Yaml and Json, which can be handled in the proposed way. And yes, you are right, this is the cause of the problem.

An alternative would be to only quote strings containing special characters, but I found no way to make the YAML::EMITTER do this only where needed.

Reply:
Thanks @wbpcode!
Reading [here](https://docs.octoprint.org/en/master/configuration/yaml.html#scalars) it seems that we may want to use double quotes by default.

It seems we use this method mainly for tests and for [runtime-entries](https://github.com/envoyproxy/envoy/blob/575cab42ec231c6e7dd5b5bad72a6ba4820e34fb/source/common/runtime/runtime_impl.cc#L238). As such, I believe the impact is pretty limited and it is safe to make the default behavior change.

Reply:
In fact, from the source code, YMAL will try only quote strings containing special characters by default. But there are some special characters it doesn't recognize.

```
StringFormat::value ComputeStringFormat(const std::string& str,
                                        EMITTER_MANIP strFormat,
                                        FlowType::value flowType,
                                        bool escapeNonAscii) {
  switch (strFormat) {
    case Auto:
      if (IsValidPlainScalar(str, flowType, escapeNonAscii)) {
        return StringFormat::Plain;
      }
      return StringFormat::DoubleQuoted;
```

Reply:
Btw, the pull request for the yaml-cpp fix is: https://github.com/jbeder/yaml-cpp/pull/1094

Reply:
Hi, thanks for your careful debug. I agree we should fix it in the `yaml-cpp`. And it's also ok to work around this problem in the Envoy.

I'm concerned that there are still some other symbols that may not be handled correctly, so it seems safer to always add quotes.

Reply:
On the yaml-cpp repo traffic seems to be low. I have added the workaround here, so that double quotes are always used. Ok?

Reply:
Well, with this change, there are now three tests failing. I could fix `test/common/protobuf/utility_test`, but with the other two I can't explain, why they fail. They read yaml only, there is no reason I see, that emitting YAML with quotes affects them.

The other two are:
```
//test/server/config_validation:server_test
//test/common/runtime:runtime_impl_test
```
I will take a deeper look tomorrow.

Reply:
The pull request on the yaml-cpp repo has been accepted and merged to their master.